net/gateway_test.go

```diff
@@ -55,73 +54,89 @@ func (t *testRandomnessServer) Home(context.Context, *drand.HomeRequest) (*drand
 
 func TestListener(t *testing.T) {
 	ctx := context.Background()
-	addr1 := "127.0.0.1:4000"
-	peer1 := &testPeer{addr1, false}
-	//addr2 := "127.0.0.1:4001"
+	hostAddr := "127.0.0.1"
+	addrGRPC := hostAddr + ":4000"
+	peerGRPC := &testPeer{addrGRPC, false}
+	addrREST := hostAddr + ":4001"
+	peerREST := &testPeer{addrREST, false}
 	randServer := &testRandomnessServer{round: 42}
 
-	lis1 := NewTCPGrpcListener(addr1, randServer)
-	go lis1.Start()
-	defer lis1.Stop()
+	lisGRPC, err := NewGRPCListenerForPublicAndProtocol(ctx, addrGRPC, randServer)
+	require.NoError(t, err)
+	lisREST, err := NewRESTListenerForPublic(ctx, addrREST, randServer)
+	require.NoError(t, err)
+	go lisGRPC.Start()
+	defer lisGRPC.Stop(ctx)
+	go lisREST.Start()
+	defer lisREST.Stop(ctx)
 	time.Sleep(100 * time.Millisecond)
 
+	// GRPC
 	client := NewGrpcClient()
-	resp, err := client.PublicRand(ctx, peer1, &drand.PublicRandRequest{})
+	resp, err := client.PublicRand(ctx, peerGRPC, &drand.PublicRandRequest{})
 	require.NoError(t, err)
 	expected := &drand.PublicRandResponse{Round: randServer.round}
 	require.Equal(t, expected.GetRound(), resp.GetRound())
 
-	err = client.PartialBeacon(ctx, peer1, &drand.PartialBeaconPacket{})
-	require.Error(t, err)
-
+	// REST
 	rest := NewRestClient()
-	resp, err = rest.PublicRand(ctx, peer1, &drand.PublicRandRequest{})
+	resp, err = rest.PublicRand(ctx, peerREST, &drand.PublicRandRequest{})
 	require.NoError(t, err)
 	expected = &drand.PublicRandResponse{Round: randServer.round}
 	require.Equal(t, expected.GetRound(), resp.GetRound())
 }
 
 // ref https://bbengfort.github.io/programmer/2017/03/03/secure-grpc.html
 func TestListenerTLS(t *testing.T) {
+	ctx := context.Background()
 	if run.GOOS == "windows" {
 		fmt.Println("Skipping TestClientTLS as operating on Windows")
 		t.Skip("crypto/x509: system root pool is not available on Windows")
 	}
-	ctx := context.Background()
-	addr1 := "127.0.0.1:4000"
-	peer1 := &testPeer{addr1, true}
+	hostAddr := "127.0.0.1"
+	addrGRPC := hostAddr + ":4000"
```

Comment:
can we use dynamic ports for these tests? golang may schedule tests to run in parallel, and if both this and the previous one run at the same time, they'll fail to listen on the same ports.

Reply:
Go does not run tests in parallel. This is part of the spec. If you want test parallelism, you explicitly orchestrate it using subtests. These tests are safe for now.

If one later wants to add a feature "parallel testing", then the right solution is not to pick a random available port, but rather to have the test orchestrator assign ports to the parallel subtests.

Picking a random test port violates two rules:
(1) It is not testing the desired code path. (There is no random port picking in the code.)
(2) It makes the test non-deterministic.

Reply:
> The ‘go test’ command may run tests for different packages in parallel up to GOMAXPROCS

Reply:
Hm, indeed.